program-test/src/programs.rs

```diff
@@ -21,76 +26,189 @@ mod spl_associated_token_account {
+    program_id: Pubkey,
+    loader_id: Pubkey,
+    elf: &'static [u8],
+}
+
+static SPL_PROGRAMS: &[ProgramInfo] = &[
+    ProgramInfo {
+        program_id: spl_token::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_token-3.5.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_token_2022::ID,
+        loader_id: bpf_loader_upgradeable::ID,
+        elf: include_bytes!("programs/spl_token_2022-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_1_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_3_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-3.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_associated_token_account::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_associated_token_account-1.1.1.so"),
+    },
 ];
 
+/// Describes one or more accounts that belong to the same program.
+///
+/// Different loaders use different account structure, so it could be useful to know which loader a
+/// particular program uses.
+pub enum ProgramAccounts {
+    /// [`bpf_loader`] aka loader v2.  Non-upgradable programs use a single account.  This account
+    /// will just directly hold ELF of the program.
+    V2 {
+        address: Pubkey,
+        account: AccountSharedData,
+    },
+
+    /// [`bpf_loader_upgradeable`] aka loader v3.  Upgradeable programs, use two accounts.
+    ///
+    /// ELF data is stored in the `program_data_account`, but it is prefixed by an additional
+    /// header.  See [`UpgradeableLoaderState::ProgramData`].
+    V3 {
+        main_address: Pubkey,
+        main_account: AccountSharedData,
+        program_data_address: Pubkey,
+        program_data_account: AccountSharedData,
+    },
+}
+
+impl From<ProgramAccounts> for Vec<(Pubkey, AccountSharedData)> {
+    fn from(accounts: ProgramAccounts) -> Self {
+        match accounts {
+            ProgramAccounts::V2 { address, account } => vec![(address, account)],
+            ProgramAccounts::V3 {
+                main_address,
+                main_account,
+                program_data_address,
+                program_data_account,
+            } => vec![
+                (main_address, main_account),
+                (program_data_address, program_data_account),
+            ],
+        }
+    }
+}
+
+pub type ProgramElf = Vec<u8>;
+
+/// Returns the program ELF and the specified program account(s).
+///
+/// For loader v2, non-upgradable programs, `elf` data will be the same as the one stored in the
+/// program account.  But for the loader v3 and loader 4, upgradable programs, the relationship is a
+/// bit more complex and it could be convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_address: Pubkey,
+) -> Result<(ProgramElf, ProgramAccounts), ProgramLoadError> {
```

Comment:
I understand that `fn spl_programs()` predates you, but since this one is a new api, I'd still prefer it did not return a tuple.

Reply:
Why?
It would help if you would provide more reasoning, as I am not sure what is your motivation here.

It is a two element tuple with types that clearly state what the elements are.
At the call site the values will be named, if the code is properly written.
I see no downsides.

If I add a struct, it will mostly add noise, I think.
Compare:

```rust
let (elf, accounts) = spl_program(rent, spl_token_2022::ID);
```

```rent
let ProgramTestProgram {
    elf,
    accounts
} = spl_program(rent, spl_token_2022::ID);
```

The latter just seems to be more verbose.

And if the caller wants to use two programs, they would need to use explicit renaming, making it even more verbose.

Should there be 3 or 4 fields, with some of the fields being optional at different call sites, then I would certainly introduce structs to name the fields.  This is what is done for things stored in `ProgramAccounts`.